### PR TITLE
New: `max-len`: `ignoreStrings`+`ignoreTemplateLiterals` (fixes #5805)

### LINKED ISSUE #5805: new rule option request: max-len ignores strings
Currently, `max-len` [supports](http://eslint.org/docs/rules/max-len) `ignoreComments`, `ignoreUrls`, etc. I'd like to request an additional option, `ignoreStrings`.

My rationale is the following:
 - strings are an atomic token, and it is a code smell to treat them as anything other than a single unbroken unit.
 - trivial changes in the string contents might necessitate the string being re-joined and re-broken, causing unnecessary diff churn
 - people use line length limits for many things, but i personally use them as a proxy for limiting line complexity - as such, length isn't actually what i care about.

This affects both inline strings as well as "require" paths and "import from" paths (and should apply to template literals as well as single/double-quoted strings)

### PR DESCRIPTION
**Pull request by ljharb**
**What issue does this pull request address?**
#5805

**What changes did you make? (Give an overview)**
 - added `getAllStrings` and `getAllTemplateLiterals` to source-code
 - grouped both by line number in `max-len`, and ignored lines that contained strings/template literals, matching the options

**Is there anything you'd like reviewers to focus on?**
The overall approach.

### CODE REVIEW

File: lib/util/source-code.js [hunk 1]
```diff
@@ -185,6 +185,26 @@ SourceCode.prototype = {
     },
 
     /**
+     * Retrieves an array containing all strings (" or ') in the source code.
```
Comment (nzakas) [inline lib/util/source-code.js]: This object is part of the public API, and I'd rather not expose these methods. Can you move these into local functions in the rule?

Comment (ljharb) [inline lib/util/source-code.js]: Sure, I'd figured it was cleaner to expose it here, but I'll move them in.

File: docs/rules/max-len.md [hunk 2]
```diff
@@ -23,6 +23,8 @@ This rule has a number or object option:
 * `"ignoreComments": true` ignores all trailing comments and comments on their own line
 * `"ignoreTrailingComments": true` ignores only trailing comments
 * `"ignoreUrls": true` ignores lines that contain a URL
+* `"ignoreStrings": true` ignores lines that contain a string (" or ')
```
Comment (platinumazure) [inline docs/rules/max-len.md]: Maybe "contain a double-quoted or single-quoted string"?

File: lib/rules/max-len.js [hunk 3]
```diff
@@ -180,6 +188,59 @@ module.exports = {
...
+         * Ensure that an array exists at [key] on `object`, and add `value` to it.
+         *
+         * @param {Object} object the object to mutate
+         * @param {string} key the object's key
+         * @param {*} value the value to add
+         * @returns {void}
+         * @private
+         */
+        function ensureArrayAndPush(object, key, value) {
+            if (!Array.isArray(object, key)) {
```
Comment (platinumazure) [inline lib/rules/max-len.js]: Should this be `Array.isArray(object[key])`? MDN doesn't mention a two-argument version.

Comment (platinumazure) [inline lib/rules/max-len.js]: Also, this suggests possible lack of test coverage in the "else case" here (there is no else, but I'm referring to this function being called twice in the same test case, where the second time should not reinitialize the array).

**Comment by platinumazure** [comment]
Raised a question about the use of `Array.isArray(obj, propName)`. Everything else LGTM.

File: docs/rules/max-len.md (hunk 2 above)
Comment (ljharb) [inline docs/rules/max-len.md]: sure, sounds good

File: lib/rules/max-len.js (hunk 3 above)
Comment (ljharb) [inline lib/rules/max-len.js]: whoops, yeah that's totally my bad, this should be `object[key]`.

**Comment by platinumazure** [comment]
Thanks! Just one more thing (and I only thought of this just now, sorry): Can we test with a line continuation somehow?

Since the code in each test case is itself a string, I think that would look something like this:

```js
"var str = \"a long string\\nwith continuation\"";
// ------------------------^ Line continuation backslash
// -------------------------^ Literal newline within the string
```

Thanks a lot!

**Comment by ljharb** [comment]
@platinumazure done

**Comment by platinumazure** [comment]
Thanks, LGTM but would like another set of eyes on this.

**Comment by ilyavolodin** [comment]
LGTM, but waiting another day for others to look

**Comment by platinumazure** [comment]
@ilyavolodin What do you think? Is this ready to go?

**Pull request merged**